### DISCUSSION TITLE
When implementing printscu, image conversion occurs cheaply

### DISCUSSION THREAD

**Discussion by jackshenjunkai** [dcm4che/dcm4che#1336] [General]
public static byte[] convertToGreyImage(File imgFile) throws IOException {
        DicomInputStream dicomInputStream = new DicomInputStream(imgFile);
//        Attributes attrs = dicomInputStream.readDataset();
        ImageInputStream imageInputStream = ImageIO.createImageInputStream(dicomInputStream);

        // 获取DICOM图像读取器
        Iterator<ImageReader> imageReaders = ImageIO.getImageReadersByFormatName("DICOM");
        ImageReader imageReader = imageReaders.next();

        // DICOM图像读取参数
        DicomImageReadParam dicomReadParam = (DicomImageReadParam) imageReader.getDefaultReadParam();
        dicomReadParam.setWindowCenter(40); // DICOM窗位
        dicomReadParam.setWindowWidth(400); // DICOM窗宽

        //自动计算窗位窗宽
        dicomReadParam.setAutoWindowing(true);
//        dicomReadParam.setPresentationState(attrs);
        dicomReadParam.setDestinationType(ImageTypeSpecifier.createGrayscale(8, DataBuffer.TYPE_BYTE,false));

        // 设置图像读取器的输入流
        imageReader.setInput(imageInputStream, false);

        // 读取DICOM图像
        BufferedImage dicomImage = imageReader.read(0, dicomReadParam);

        //获取图像偏移信息
        int offsetX = dicomImage.getMinX();
        int offsetY = dicomImage.getMinY();

        // 将DICOM图像转换为灰度图像
        BufferedImage greyImage = new BufferedImage(dicomImage.getWidth(), dicomImage.getHeight(), BufferedImage.TYPE_BYTE_GRAY);
        Graphics2D graphics = greyImage.createGraphics();
        graphics.drawImage(dicomImage, -offsetX, -offsetY, null);
        graphics.dispose();

        // 获取TIFF ImageWriter
        ImageWriter writer = ImageIO.getImageWritersByFormatName("TIF").next();

        // 设置压缩参数（可选）
        ImageWriteParam param = writer.getDefaultWriteParam();
        param.setCompressionMode(ImageWriteParam.MODE_DEFAULT);

//         将灰度图像转换为字节数组
        // 将图像写入 ByteArrayOutputStream
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        ImageOutputStream ios = ImageIO.createImageOutputStream(outputStream);
        writer.setOutput(ios);
        writer.write(null, new javax.imageio.IIOImage(greyImage, null, null), param);
        ios.close();
//        ImageIO.write(greyImage, "TIFF", outputStream);
        byte[] greyBytes = outputStream.toByteArray();

        // 关闭流
        dicomInputStream.close();
        imageInputStream.close();
        outputStream.close();

        return greyBytes;
    }

![image](https://github.com/dcm4che/dcm4che/assets/17916158/e02812ad-142d-4e80-9c55-1771beab94d5)

**Comment by CccTAu** [comment]
hello,my bro. Do you solve your problem? I do not know how to implement printscu now. Can I get your help? Tank you very much.
